# Jupyter

Usaremos o Jupyter para todos os nossos exemplos -- isso nos permite rodar o Python em um notebook baseado na web, mantendo um histórico de entradas e saídas, juntamente com texto e imagens.

Para ajuda com o Jupyter, visite: https://jupyter.readthedocs.io/en/latest/content-quickstart.html

Nós interagimos com o Python digitando nas _células_ do notebook. Por padrão, uma célula é uma célula de _código_, o que significa que você pode inserir qualquer código Python válido nela e executá-lo. Outro tipo importante de célula é a célula de _markdown_. Isso permite que você coloque texto, com diferentes formatações (itálico, negrito, etc.), que descrevem o que o notebook está fazendo.

Você pode alterar o tipo de célula através do menu na parte superior, ou usando os atalhos:

  * ctrl-m m : célula de markdown
  * ctrl-m y : célula de código


Alguns atalhos úteis:

 * shift+enter = executar a célula e pular para a próxima (criando uma nova célula se não houver outra nova)
 * ctrl+enter = executar a célula no local
 * alt+enter = executar a célula e inserir uma nova abaixo

ctrl+m h lista outros comandos

Uma "célula de markdown" permite que você insira equações em LaTeX diretamente no seu notebook. Basta colocá-las em `$` ou `$$`:

$$\frac{\partial \rho}{\partial t} + \nabla \cdot (\rho U) = 0$$


<div class="alert alert-block alert-danger">
    
**Importante**: quando você trabalha em um notebook, tudo o que foi feito nas células anteriores ainda está na memória e é _conhecido_ pelo Python, então você pode se referir a funções e variáveis que foram definidas anteriormente. Mesmo que você vá para o topo do notebook e insira uma célula, toda a informação feita anteriormente na sessão do notebook ainda está definida -- não importa onde fisicamente você esteja no notebook. Se você quiser resetar as coisas, pode usar as opções no menu _Kernel_.
</div>

<div class="alert alert-block alert-warning">
    
<span class="fa fa-flash"></span> Exercício Rápido:

Crie uma nova célula abaixo desta. Certifique-se de que seja uma célula _de código_ e insira o seguinte código e execute-o:

    
```

 print("Hello, World")
 
```
</div>

`print()` é uma _função_ em Python que recebe argumentos (dentro dos `()`) e exibe a saída na tela. Você pode imprimir várias quantidades de uma vez, como:

In [3]:
print(1, 2, 3)

1 2 3


In [4]:
1

1

Observe que o comportamento padrão no Jupyter é imprimir o valor de retorno da última instrução em uma célula, então não precisamos usar o `print` se quisermos apenas o valor de algo como:

In [5]:
a = 10
a

10

# Tipos Básicos de Variáveis

Python é uma linguagem de tipagem dinâmica - isso significa que você não
precisa especificar com antecedência que tipo de dado você vai armazenar
em uma variável.  No entanto, existem alguns tipos de dados principais
com os quais precisamos nos familiarizar ao usar a linguagem.

O primeiro conjunto de tipos de dados é semelhante àqueles encontrados em outras
linguagens (como C/C++ e Fortran): números de ponto flutuante, inteiros
e strings.

O próximo conjunto de tipos de dados são containers.  Em Python, ao contrário de algumas
linguagens, estes são integrados à linguagem e tornam muito fácil realizar
operações complexas.  Vamos ver isso mais adiante.

Alguns exemplos vêm do tutorial oficial do Python:
http://docs.python.org/3/tutorial/

## Números Inteiros

Números inteiros são números sem um ponto decimal. Eles podem ser positivos ou negativos. A maioria das linguagens de programação usa uma quantidade finita de memória para armazenar um único inteiro, mas em Python, a quantidade de memória será expandida conforme necessário para armazenar inteiros grandes.

Os operadores básicos, `+`, `-`, `*` e `/` funcionam com inteiros.

In [ ]:
2+2+3

7

In [ ]:
2*-4

-8

```{note}
A divisão inteira é um ponto onde o Python 2 e o Python 3 são diferentes

No Python 3.x, dividir 2 inteiros resulta em um número de ponto flutuante.  No Python 2.x, dividir 2 inteiros resulta em um número inteiro.  O último é consistente com muitas linguagens de programação fortemente tipadas (como Fortran ou C), já que o tipo de dado do resultado é o mesmo das entradas, mas o primeiro está mais alinhado com as nossas expectativas.
```

In [ ]:
3/2

1.5

Para obter um resultado inteiro, usamos o operador //

In [ ]:
3//2

1

Python é uma _linguagem de tipagem dinâmica_—isso significa que não precisamos declarar o tipo de dado de uma variável antes de inicializá-la.  

Aqui vamos criar uma variável (pense nela como um rótulo descritivo que pode se referir a algum dado). O operador `=` atribui um valor à variável.

In [ ]:
a = 1.5
b = 3

Funções operam sobre variáveis e retornam um resultado. Aqui, `print()` irá exibir a saída na tela.

In [ ]:
a + b

4.5

In [ ]:
a * b

4.5

Observe que os nomes de variáveis diferenciam maiúsculas de minúsculas, então `a` e `A` são diferentes.

In [ ]:
A = 2025

In [ ]:
print(a, A)

1.5 2025


Aqui inicializamos 3 variáveis, todas com o valor `0`, mas essas ainda são variáveis distintas, então podemos mudar uma sem afetar as outras.

In [ ]:
x = y = z = 0

In [ ]:
print(x, y, z)

0 0 0


In [ ]:
z = 1

In [ ]:
z

1

O Python tem alguma ajuda embutida (e o Jupyter/IPython tem ainda mais).

Tente fazer:
```
help(x)
```

In [ ]:
help(z)

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |

Outra função, `type()`, retorna o tipo de dado de uma variável

In [ ]:
type(x)

int

```{note}
Observe que em linguagens como Fortran e C, você especifica a quantidade de memória que um inteiro pode ocupar (geralmente 2 ou 4 bytes). Isso impõe uma restrição sobre o maior tamanho de inteiro que pode ser representado. O Python adaptará o tamanho do inteiro para que você não tenha *overflow*.
```

In [ ]:
a = 12345678901234567890123456789012345123456789012345678901234567890
print(a)
print(a.bit_length())
print(type(a))

12345678901234567890123456789012345123456789012345678901234567890
213
<class 'int'>


## Ponto Flutuante

Ao operar com números de ponto flutuante e inteiros, o resultado é promovido a um float.

In [ ]:
1. + 2

3.0

Mas observe o operador especial de divisão inteira.

In [ ]:
1.//2

0.0

```{important}
É importante entender que, como existem infinitos números reais entre quaisquer dois limites, em um computador precisamos aproximar isso por um número finito. Existe um padrão IEEE para ponto flutuante que praticamente todas as linguagens e processadores seguem.

Isso significa duas coisas:

* Nem todo número real terá uma representação exata em ponto flutuante.
* Há uma precisão finita para os números — abaixo disso, perdemos o controle das diferenças (isso é geralmente chamado de *erro de arredondamento*).
```

```{tip}
O ponto flutuante é essencial para a ciência computacional.  Uma ótima
introdução ao ponto flutuante e suas limitações é: [O que todo
cientista da computação deveria saber sobre aritmética de ponto flutuante](http://dl.acm.org/citation.cfm?id=103163) de
D. Goldberg. Este artigo é uma referência incrível para entender como um computador armazena números que também pode ser encontrado nesse [link](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)
```

Considere a seguinte expressão, por exemplo:

In [ ]:
0.3/0.1 - 3

-4.440892098500626e-16

In [ ]:
0.3/0.1 == 3

False

Aqui está outro exemplo: O número `0.1` não pode ser representado exatamente em um computador. Em nossa impressão, usamos um especificador de formato (as coisas dentro de `{}`) para pedir que mais precisão seja mostrada:

In [ ]:
a = 0.1
print("{:30.20}".format(a))

        0.10000000000000000555


Podemos pedir ao Python para relatar os limites do ponto flutuante.

In [ ]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

Observe que isso diz que só podemos armazenar números entre `2.2250738585072014e-308` e `1.7976931348623157e+308`.

Também vemos que a precisão é `2.220446049250313e-16` (isso é comumente chamado de _epsilon da máquina_). Para ver isso, considere adicionar um número pequeno a `1.0`. Usaremos o operador de igualdade (`==`) para testar se dois números são iguais:

```{admonition} Exercício Rapído

Defina duas variáveis, \( a = 1 \) e \( e = 10^{-16} \).

Agora defina uma terceira variável, `b = a + e`.

Podemos usar o operador `==` do Python para testar a igualdade. O que você espera que `b == a` retorne? Execute e veja se concorda com seu palpite.
```

## Módulos e Namespaces

A linguagem Python básica é bastante pequena. A maior parte acontece em módulos. Alguns módulos fazem parte de uma biblioteca padrão que fornece funcionalidade adicional. Essas partes adicionadas estão na forma de módulos que podemos _importar_ para nossa sessão (ou programa) Python.

O módulo `math` fornece funções que realizam as operações matemáticas básicas, bem como constantes (note que há um módulo separado `cmath` para números complexos).

Em Python, você `importa` um módulo. As funções são então definidas em um _namespace_ separado — esta é uma região separada que define nomes e variáveis, etc. Uma variável em um namespace pode ter o mesmo nome que uma variável em um namespace diferente, e elas não entram em conflito. Você usa o operador `.` para acessar um membro de um namespace.

Por padrão, quando você digita algo no interpretador Python ou aqui no notebook Jupyter, ou em um script, ele está em seu próprio namespace padrão, e você não precisa prefixar nenhuma das variáveis com um indicador de namespace.

In [ ]:
import math

O módulo `math` fornece o valor de π (pi).

In [ ]:
math.pi

3.141592653589793

Isso é distinto de qualquer variável `pi` que possamos definir aqui.

In [ ]:
pi = 3

In [ ]:
print(pi, math.pi)

3 3.141592653589793


Observe que `pi` e `math.pi` são distintos um do outro—eles estão em namespaces diferentes.

### Operações de Ponto Flutuante

Os mesmos operadores, `+`, `-`, `*`, `/` funcionam como de costume para números de ponto flutuante. Para elevar um número a uma potência, usamos o operador `**` (isso é o mesmo que em Fortran).

In [ ]:
R = 2.0

In [ ]:
math.pi * R**2

12.566370614359172

A precedência dos operadores segue a de muitas linguagens. Veja [Precedência de Operadores do Python](https://docs.python.org/3/reference/expressions.html#operator-precedence) para mais detalhes.

Em ordem de precedência:
* Quantidades em `()`
* Fatiamento, chamadas, subscritos
* Exponenciação (`**`)
* `+x`, `-x`, `~x`
* `*`, `@`, `/`, `//`, `%`
* `+`, `-`

(Depois disso estão as operações bitwise e comparações)

Os parênteses podem ser usados para sobrepor a precedência.

```{admonition} Exercício Rapído
Considere as seguintes expressões. Usando as ideias de precedência, pense sobre qual valor resultará e, em seguida, experimente no código abaixo para ver se você estava certo.

* `1 + 3*2**2`
* `1 + (3*2)**2`
* `2**3**2`
```

O módulo `math` fornece muitas das funções matemáticas padrão. A maioria delas é na verdade repetida no `numpy`, então, na prática, eu pessoalmente quase nunca uso `math`.

Para as funções trigonométricas, a expectativa é que o argumento da função esteja em radianos—você pode usar `math.radians()` para converter de graus para radianos, por exemplo:

In [ ]:
math.cos(math.radians(45))

0.7071067811865476

Observe que nessa declaração estamos alimentando a saída de uma função (`math.radians()`) em uma segunda função, `math.cos()`

Quando em dúvida, peça ajuda para descobrir todas as coisas que um módulo fornece:

In [ ]:
help(math.sin)

Help on built-in function sin in module math:

sin(x, /)
    Return the sine of x (measured in radians).



## Números Complexos

O Python usa '`j`' para denotar a unidade imaginária.

In [ ]:
1.0 + 2j

(1+2j)

In [96]:
a = 1j
b = 3.0 + 2.0j
print(a + b)
print(a * b)

(3+3j)
(-2+3j)


Podemos usar `abs()` para obter a magnitude e, separadamente, obter as partes real ou imaginária.

In [97]:
print("magnitude: ", abs(b))
print("parte real: ", a.real)
print("parte imaginária: ", a.imag)

magnitude:  3.605551275463989
parte real:  0.0
parte imaginária:  1.0


## Strings

O Python não se importa se você usa aspas simples ou duplas para strings:

In [ ]:
a = "está é a minha string"
b = 'outra string'

In [ ]:
print(a)
print(b)

está é a minha string
outra string


Muitos dos operadores matemáticos usuais também são definidos para strings. Por exemplo, para concatenar ou duplicar:

In [ ]:
a + b

'está é a minha stringoutra string'

In [ ]:
a + ". " + b

'está é a minha string. outra string'

In [ ]:
a * 3

'está é a minha stringestá é a minha stringestá é a minha string'

Existem vários códigos de escape que são interpretados em strings. Esses começam com uma barra invertida, `\`. Por exemplo, você pode usar `\n` para uma nova linha.

In [ ]:
a = a + "\n" + "hello"
print(a)

está é a minha string
hello


```{admonition} Exercício Rapído
    
A função `input()` pode ser usada para pedir a entrada do usuário.

* Use `help(input)` para ver como funciona.  
* Escreva um código para pedir a entrada e armazenar o resultado em uma variável. `input()` retornará uma string.

* Use a função `float()` para converter um número inserido como entrada em uma variável de ponto flutuante.  
* Verifique se a conversão funcionou usando a função `type()`.
```

Aspas triplas `"""` podem envolver strings de várias linhas. Isso é útil para docstrings no início das funções (mais sobre isso mais tarde...).

In [ ]:
c = """
É um período de guerra civil. Naves rebeldes, 
atacando de uma base oculta, conquistaram sua primeira 
vitória contra o maligno Império Galáctico.

Durante a batalha, espiões rebeldes conseguiram roubar 
os planos secretos da arma definitiva do Império, a 
Estrela da Morte, uma estação espacial blindada com poder 
suficiente para destruir um planeta inteiro.

Perseguida pelos sinistros agentes do Império, a Princesa 
Leia corre para casa a bordo de sua nave estelar, guardiã 
dos planos roubados que podem salvar seu povo e 
restaurar a liberdade na galáxia..."""

In [ ]:
print(c)


É um período de guerra civil. Naves rebeldes, 
atacando de uma base oculta, conquistaram sua primeira 
vitória contra o maligno Império Galáctico.

Durante a batalha, espiões rebeldes conseguiram roubar 
os planos secretos da arma definitiva do Império, a 
Estrela da Morte, uma estação espacial blindada com poder 
suficiente para destruir um planeta inteiro.

Perseguida pelos sinistros agentes do Império, a Princesa 
Leia corre para casa a bordo de sua nave estelar, guardiã 
dos planos roubados que podem salvar seu povo e 
restaurar a liberdade na galáxia...


Uma string bruta não substitui sequências de escape (como `\n`). Basta colocar um `r` antes da primeira aspa:

In [ ]:
d = r"está é uma string bruta \n hello"
d

'está é uma string bruta \\n hello'

O fatiamento é usado para acessar uma parte de uma string.

Fatiar uma string pode parecer um pouco contra-intuitivo se você vem de C ou Fortran. O truque é pensar no índice como representando a borda esquerda de um caractere na string. Quando fizermos arrays mais tarde, o mesmo se aplicará.

Observe também que o Python (como C) usa indexação baseada em 0.

Os índices negativos contam a partir da direita.

In [ ]:
# Definindo uma string de exemplo
a = "está é a minha string"

# Imprimindo a string completa
print(a) 

# Imprimindo uma parte da string usando fatiamento
# a[11:14] significa: comece no índice 11 e vá até o índice 14 (exclusivo)
print(a[11:14]) 

# Imprimindo o primeiro caractere da string
# a[0] significa: pegue o caractere no índice 0
print(a[0])


# Imprimindo o segundo caractere a partir do final da string 'a'
print(a[-2])

# Imprimindo uma parte da string usando fatiamento com passo
# a[11:20:2] significa: comece no índice 11, vá até o índice 20 (exclusivo), pegando a cada 2 caracteres
print(a[11:20:2])  # Saída: "mni" (caracteres nos índices 11, 13, 15, 17, 19)

está é a minha string
nha
e
n
nasrn


```{admonition} Quick Exercise

As strings têm muitos _métodos_ (funções que sabem como trabalhar com um tipo de dado específico, neste caso, strings). Um método útil é `.find()`. Para uma string `a`, `a.find(s)` retornará o índice da primeira ocorrência de `s`.

Para nossa string `c` acima, encontre o primeiro `.` (identificando a primeira frase completa) e imprima apenas a primeira frase em `c` usando esse resultado.

```

Existem também vários métodos e funções que trabalham com strings. Aqui estão alguns exemplos:

In [ ]:
print(a)  # Exibe a string original

print(a.replace("está", "aquela"))  # Exibe a string com "esta" substituído por "aquela"

print(len(a))  # Exibe o número total de caracteres na string

print(a)  # Exibe a string original novamente, sem alterações

está é a minha string
aquela é a minha string
21
está é a minha string


Note que nossa string original, `a`, não foi alterada. Em Python, as strings são *imutáveis*. Operações em strings retornam uma nova string.

In [ ]:
type(a)

str

Podemos formatar strings ao imprimir para inserir quantidades em lugares específicos na string. Um `{}` serve como um espaço reservado para uma quantidade e é substituído usando o método `.format()`:

In [98]:
a = 1
b = 2.0
c = "teste"
print("a = {}; b = {}; c = {}".format(a, b, c))

a = 1; b = 2.0; c = teste


Poderiamos juntar string também usando `+`

In [ ]:
print("a="+str(a))

a=1


Mas a maneira mais moderna de fazer isso é usar *f-strings*.

In [99]:
print(f"a = {a}; b = {b}; c = {c}")

a = 1; b = 2.0; c = teste


Note o `f` precedendo a aspa inicial `"` 

Uma outra opção é a seguinte, onde você pode especificar o formato de cada variável:

In [100]:
print("a = %i; b = %.2f; c = %s" % (a, b, c))

a = 1; b = 2.00; c = teste


# Variáveis de Tipo Avançado

Estas notas seguem o tutorial oficial do Python de forma bastante próxima: http://docs.python.org/3/tutorial/

## Listas

Listas agrupam dados. Muitas linguagens têm arrays (vamos olhar para eles em breve no Python). Mas, ao contrário dos arrays na maioria das linguagens, as listas podem conter dados de todos os tipos diferentes — não precisam ser homogêneas. Os dados podem ser uma mistura de inteiros, números de ponto flutuante ou complexos, strings ou outros objetos (incluindo outras listas).

Uma lista é definida usando colchetes:

In [101]:
a = [1, 2.0, "minha lista", 4]

In [102]:
a

[1, 2.0, 'minha lista', 4]

Podemos indexar uma lista para obter um único elemento — lembre-se de que o Python começa a contagem em 0:

In [103]:
a[2]

'minha lista'

In [105]:
a+["hello"]

[1, 2.0, 'minha lista', 4, 'hello']

Assim como com strings, operadores matemáticos são definidos em listas:

In [104]:
a*2

[1, 2.0, 'minha lista', 4, 1, 2.0, 'minha lista', 4]

A função `len()` retorna o comprimento de uma lista.

In [60]:
len(a)

4

Ao contrário das strings, as listas são _mutáveis_ — você pode alterar elementos em uma lista facilmente.

In [106]:
print(a)
a[1] = -2.0
print(a)

[1, 2.0, 'minha lista', 4]
[1, -2.0, 'minha lista', 4]


In [107]:
a

[1, -2.0, 'minha lista', 4]

In [108]:
a[0:1] = [-1, -2.1]   # isso colocará dois itens no lugar onde 1 existia antes
a

[-1, -2.1, -2.0, 'minha lista', 4]

Note que as listas podem até conter outras listas:

In [109]:
a[1] = ["outra lista", 3]
a

[-1, ['outra lista', 3], -2.0, 'minha lista', 4]

Assim como tudo no Python, uma lista é um objeto que é uma instância de uma classe. As classes têm métodos (funções) que sabem como operar em um objeto daquela classe.

Existem muitos métodos que funcionam em listas. Dois dos mais úteis são `append`, para adicionar ao final de uma lista, e `pop`, para remover o último elemento:

In [110]:
a.append(6)
a

[-1, ['outra lista', 3], -2.0, 'minha lista', 4, 6]

In [111]:
a.pop()

6

In [112]:
a

[-1, ['outra lista', 3], -2.0, 'minha lista', 4]

```{admonition} Exercício Rápido:

Uma operação que veremos muito é começar com uma lista vazia e adicionar elementos a ela. Uma lista vazia é criada como:

 a = []

* Crie uma lista vazia
* Adicione os inteiros de 1 a 10 a ela.
* Agora remova-os da lista um por um.
  
```

### Copiando listas

Copiar pode parecer um pouco contra-intuitivo no início. A melhor maneira de pensar sobre isso é que sua lista vive em algum lugar na memória e, quando você faz 

```
a = [1, 2, 3, 4]
```

então a variável `a` é configurada para apontar para aquele local na memória, então ela se refere à lista.

Se então fizermos
```
b = a
```
então `b` também apontará para aquele mesmo local na memória — o exato mesmo objeto de lista.

Como ambos estão apontando para o mesmo local na memória, se alterarmos a lista através de `a`, a mudança será refletida em `b` também:

In [113]:
a = [1, 2, 3, 4]
b = a  # tanto a quanto b se referem ao mesmo objeto de lista na memória
print(a)
a[0] = "alteração"
print(b)

[1, 2, 3, 4]
['alteração', 2, 3, 4]


Se você quiser criar um novo objeto na memória que seja uma cópia de outro, você pode indexar a lista, usando `:` para obter todos os elementos, ou usar a função `list()`:

In [114]:
c = list(a)   # você também pode fazer c = a[:], que basicamente fatiará toda a lista
a[1] = "dois"
print(a)
print(c)

['alteração', 'dois', 3, 4]
['alteração', 2, 3, 4]


As coisas ficam um pouco complicadas quando uma lista contém outro objeto mutável, como outra lista. Então, a cópia que analisamos acima é apenas uma _cópia rasa_. Faremos isso com mais cuidado na próxima vez.

Quando estiver em dúvida, use a função `id()` para descobrir onde na memória um objeto está localizado (você não deve se preocupar com o que os valores dos números que você obtém de `id` significam, mas apenas se eles são os mesmos que os de outro objeto).

In [75]:
print(id(a), id(b), id(c))

4436542400 4436542400 4430978048


ou use o operador `is`

In [76]:
a is b

True

In [77]:
a is c

False

Existem muitos outros métodos que funcionam em listas

In [78]:
my_list = [10, -1, 5, 24, 2, -1, 9]
my_list.sort() # Ordena a lista em ordem crescente
my_list

[-1, -1, 2, 5, 9, 10, 24]

In [79]:
my_list.count(-1) # Conta quantas vezes o valor -1 aparece na lista

2

Podemos também inserir elementos.

In [115]:
a.insert(3, "meu elemento inserido")
a

['alteração', 'dois', 3, 'meu elemento inserido', 4]

Juntar duas listas é simples. Assim como com strings, o operador `+` concatena:

In [81]:
b = [1, 2, 3]
c = [4, 5, 6]
d = b + c
print(d)

[1, 2, 3, 4, 5, 6]


## Dicionários

Um dicionário armazena dados como um par `chave:valor`. Ao contrário de uma lista, onde você tem uma ordem específica, as chaves em um dicionário permitem que você acesse informações facilmente de qualquer lugar:

In [53]:
my_dict = {"chave1":1, "chave2":2, "chave3":3}

In [54]:
my_dict["chave1"]

1

Você pode adicionar uma nova `chave:valor` facilmente, e ela pode ser de qualquer tipo.

In [55]:
my_dict["novachave"] = "nova"
my_dict

{'chave1': 1, 'chave2': 2, 'chave3': 3, 'novachave': 'nova'}

Você também pode obter facilmente a lista de chaves que estão definidas em um dicionário.

In [56]:
chaves = list(my_dict.keys())
chaves

['chave1', 'chave2', 'chave3', 'novachave']

e verificar facilmente se uma chave existe no dicionário usando o operador `in`.

In [57]:
print("chave1" in chaves)
print("ChaveInvalida" in chaves)

True
False


```{admonition} Exercício Rapído

Crie um dicionário onde as chaves são os nomes em string dos números de zero a nove e os valores são suas representações numéricas (0, 1, ... , 9).

```

```{note}
:class: dropdown
## Solução
```python
# Criando um dicionário com os nomes dos números de zero a nove como chaves
number_dict = {
    "zero": 0,
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9
}

# Exibindo o dicionário
print(number_dict)
```


## Compreensões de Lista

As compreensões de lista fornecem uma maneira compacta de inicializar listas. Alguns exemplos do tutorial:

In [88]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [89]:
squares = [x**2 for x in range(10)]

In [90]:
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

Aqui, usamos outro tipo de Python, a tupla, para combinar números de duas listas em um par.

In [91]:
[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]

[(1, 3), (1, 4), (2, 3), (2, 1), (2, 4), (3, 1), (3, 4)]

```{admonition} Exercício Rapído

Use uma compreensão de lista para criar uma nova lista a partir de `squares` contendo apenas os números pares. Pode ser útil usar o operador de módulo, `%`.

```

## Tuplas

As tuplas são imutáveis — elas não podem ser alteradas, mas são úteis para organizar dados em algumas situações. Usamos () para indicar uma tupla:

In [64]:
a = (1, 2, 3, 4)
a

(1, 2, 3, 4)

Podemos desempacotar uma tupla:

In [94]:
w, x, y, z = a

In [95]:
w

1

Como uma tupla é imutável, não podemos alterar um elemento:

In [96]:
a[0] = 2

TypeError: 'tuple' object does not support item assignment

Mas podemos transformá-la em uma lista, e então podemos alterá-la.

In [67]:
z = list(a)

In [68]:
z[0] = "novo"

In [69]:
z

['novo', 2, 3, 4]

Frequentemente, não está claro como as tuplas diferem das listas. A diferença mais óbvia é que elas são imutáveis (use isso a seu favor para evitar bugs!). Muitas vezes, veremos tuplas usadas para armazenar dados relacionados que devem ser interpretados juntos. Um bom exemplo é um ponto cartesiano, (x, y). Aqui está uma lista de pontos:

In [100]:
points = []
points.append((1,2))
points.append((2,3))
points.append((3,4))
points

[(1, 2), (2, 3), (3, 4)]

Podemos até gerar esses para uma curva usando uma compreensão de lista:

In [101]:
points = [(x, 2*x + 5) for x in range(10)]
points

[(0, 5),
 (1, 7),
 (2, 9),
 (3, 11),
 (4, 13),
 (5, 15),
 (6, 17),
 (7, 19),
 (8, 21),
 (9, 23)]

# Fluxo de Controle

Estas notas seguem o tutorial oficial do Python de forma bastante próxima: http://docs.python.org/3/tutorial/

Para escrever um programa, precisamos da capacidade de iterar e tomar ações com base nos valores de uma variável. Isso inclui testes condicionais (if-tests) e loops.

**O Python usa 4 espaços em branco para denotar um bloco de código**. A maioria dos editores interpretará uma tabulação como 4 espaços em branco por padrão (se não, verifique suas configurações).

## Loops

Um simples laço while — note a indentação para denotar o bloco que faz parte do laço.

Aqui também usamos o operador compacto `+=`: `n += 1` é o mesmo que `n = n + 1`.

In [102]:
n = 0
while n < 10:
    print(n)
    n += 1

0
1
2
3
4
5
6
7
8
9


Este foi um exemplo muito simples. Mas frequentemente usaremos a função `range()` nesta situação. Note que `range()` pode receber um passo (stride).

In [104]:
list(range(2, 10, 2))

[2, 4, 6, 8]

Vamos analisar cada parte:

- **`range(start, stop, step)`**: A função `range()` gera uma sequência de números. Ela aceita três argumentos:
  - **`start`**: O valor inicial da sequência (inclusivo). No nosso caso, é `2`.
  - **`stop`**: O valor final da sequência (exclusivo). Aqui, é `10`, o que significa que a sequência irá parar antes de chegar a `10`.
  - **`step`**: O incremento entre cada número na sequência. Neste exemplo, o passo é `2`, o que significa que os números gerados aumentarão de `2` em `2`.

In [103]:
for n in range(2, 10, 2):
    print(n)

2
4
6
8


## Instruções If

`if` permite a ramificação. O Python não possui uma instrução select/case como algumas outras linguagens, mas `if`, `elif` e `else` podem reproduzir qualquer funcionalidade de ramificação que você possa precisar.

In [116]:
x = 0

if x < 0:
    print("negativo")
elif x == 0:
    print("zero")
else:
    print("positivo")


zero


## Iterando sobre elementos

É fácil iterar sobre itens em uma lista ou qualquer objeto _iterável_. O operador `in` é a chave aqui.

In [117]:
alist = [1, 2.0, "três", 4]
for a in alist:
    print(a)

1
2.0
três
4


In [118]:
## Don't do this
for i in [0,1,2,3]:
    print(alist[i])

1
2.0
três
4


In [70]:
for c in "isto é uma string":
    print(c)

i
s
t
o
 
é
 
u
m
a
 
s
t
r
i
n
g


Podemos combinar laços e testes `if` para realizar lógicas mais complexas, como sair do laço quando você encontra o que está procurando.

In [119]:
n = 0
for a in alist:
    if a == "três":
        break
    else:
        n += 1

print(n)


2


(No entanto, para esse exemplo, há uma maneira mais simples.)

In [120]:
alist.index("três")

2

Para dicionários, você também pode iterar sobre os elementos.

In [71]:
my_dict = {"chave1":1, "chave2":2, "chave3":3}

for k, v in my_dict.items():
    print("chave = {}, valor = {}".format(k, v))    # notice how we do the formatting here


chave = chave1, valor = 1
chave = chave2, valor = 2
chave = chave3, valor = 3


In [72]:
for k in my_dict:
    print(k, my_dict[k])

chave1 1
chave2 2
chave3 3


Às vezes, queremos iterar sobre um elemento de uma lista e saber seu índice — `enumerate()` ajuda aqui:

In [121]:
for n, a in enumerate(alist):
    print(n, a)

0 1
1 2.0
2 três
3 4


   
A função `zip()` nos permite iterar sobre dois iteráveis ao mesmo tempo. Assim `zip(a, b)` atuará como uma lista, onde cada elemento é uma tupla com um item de `a` e o elemento correspondente de `b`.

In [116]:
a = [1, 2, 3, 4, 5, 6, 7, 8]
b = ["a", "b", "c", "d", "e", "f", "g", "h"]


for x,y in zip(a,b):
    print(x,y)



1 a
2 b
3 c
4 d
5 e
6 f
7 g
8 h


In [117]:
list(zip(a,b))

[(1, 'a'),
 (2, 'b'),
 (3, 'c'),
 (4, 'd'),
 (5, 'e'),
 (6, 'f'),
 (7, 'g'),
 (8, 'h')]

```{admonition} Exercício Rápido:
    

A função `.split()` em uma string pode dividi-la em palavras (separando por espaços).

Usando `.split()`, itere sobre as palavras na string

`a = "O rato roeu a roupa do rei de roma"`

e imprima uma palavra por linha.

```

# Funções

As funções são usadas para organizar o fluxo do programa, especialmente para nos permitir realizar facilmente tarefas comumente necessárias repetidamente. Já usamos muitas funções, como aquelas que trabalham com listas (`append()` e `pop()`) ou strings (como `replace()`). Aqui, veremos como escrever nossas próprias funções.

Uma função recebe argumentos, listados entre `()`, e retorna um valor. Mesmo que você não forneça explicitamente um valor de retorno, um valor será retornado (por exemplo, `None`).

Aqui está um exemplo simples de uma função que recebe um único argumento, `i`:

In [80]:
def minha_func(i):
    print(f"na função temos que i = {i}")
    
minha_func(10)
minha_func(5)

na função temos que i = 10
na função temos que i = 5


As funções são um lugar onde o _escopo_ entra em jogo. Uma função tem seu próprio _namespace_. Se uma variável não estiver definida nessa função, ela procurará no namespace de onde foi chamada para ver se essa variável existe lá.

No entanto, você deve evitar isso tanto quanto possível (variáveis que persistem entre namespaces são chamadas de variáveis globais).

Veja se você entende o que está acontecendo aqui:

In [81]:
c=100

def soma_c(a):
    c= 10
    return a+c

def soma_c_denovo(a):
    return a+c

soma_c(1), soma_c_denovo(1)



(11, 101)

As funções sempre retornam um valor — se um valor não for explicitamente fornecido, então elas retornam `None`; caso contrário, podem retornar valores (até mesmo múltiplos valores) de qualquer tipo.

In [82]:
a = minha_func(10)

na função temos que i = 10


In [77]:
print(a)

None


Aqui está uma função simples que recebe dois números e retorna seu produto.

In [83]:
def multiplique(a, b):
    return a*b

c = multiplique(3, 4)
c

12

Retornar mais números é simples (sem estruturas, ponteiros e tudo isso).

In [84]:
def medêambos(a, b):
    return a+10, b+10

medêambos(1,2)

(11, 12)

```{admonition} Exercício Rapído
    
Escreva uma função simples que recebe uma frase (como uma string) e retorna um inteiro igual ao comprimento da palavra mais longa na frase. A função `len()` e o método `.split()` serão úteis aqui.

```

`None` é uma quantidade especial em Python (análoga ao `null` em algumas outras linguagens). Podemos testar se um valor é `None` — a maneira preferida é usar `is`:

In [85]:
def faça_nada():
    pass

a = faça_nada()
if a is None:
    print("não fizemos nada")

não fizemos nada


In [132]:
a is None

True

## Funções Mais Complexas

Aqui está um exemplo mais complexo em que retornamos um par de variáveis. Também note a _docstring_ aqui.

```{information}
A sequência de Fibonacci é uma série de números inteiros que começa com 0 e 1, na qual cada número seguinte é a soma dos dois números anteriores.
```

In [90]:
def fib2(n):
    """Retorna uma lista contendo a série de Fibonacci até n."""
    result = []
    a, b = 0, 1
    while a < n:
        result.append(a)    # see below
        a, b = b, a+b
    return result, len(result)

fib, n = fib2(250)
print(fib,"\n",n)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233] 
 14


Observe que esta função inclui uma docstring (logo após a definição da função). Ela é utilizada pela função `help`.

In [91]:
help(fib2)

Help on function fib2 in module __main__:

fib2(n)
    Retorna uma lista contendo a série de Fibonacci até n.



Você pode ter argumentos opcionais que fornecem valores padrão. Aqui está uma função simples que valida uma resposta, com um argumento opcional que pode fornecer a resposta correta.

In [122]:
def checar_resposta(val, resposta_correta="a"):
    return val == resposta_correta

print(checar_resposta("a"))
print(checar_resposta("a", resposta_correta="b"))

True
False


```{admonition} Exercício
    

Execute esses dois blocos de código e entenda/explique a diferença... **Isso leva a um dos erros mais comuns para iniciantes em Python**

Isto:

````python    
    def f(a, L=[]):
        L.append(a)
        return L

    print(f(1))
    print(f(2))
    print(f(3))
    
e isto:
 ````python    
    def fnew(a, L=None):
        if L is None:
            L = []
        L.append(a)
        return L

    print(fnew(1))
    print(fnew(2))
    print(fnew(3))   

```

```{tip} 
:class: dropdown
### Solução
Observe que cada chamada na função `f` não cria sua própria lista separada. Em vez disso, uma única lista vazia foi criada quando a função `f` foi processada pela primeira vez, e essa lista persiste na memória como o valor padrão para o argumento opcional `L`.

Se quisermos que uma lista única seja criada a cada vez (ou seja, um local separado na memória), devemos inicializar o valor do argumento como None e, em seguida, verificar seu valor real e criar uma lista vazia no corpo da função se o valor padrão não tiver sido alterado. Assim como foi executado na função `fnovo`
```

## Lambdas

Lambdas são funções "descartáveis". Essas são pequenas funções sem nome que são frequentemente usadas como argumentos em outras funções. Os seguintes exemplos são equivalentes:

In [80]:
def aoquadrado(x):
    return x**2

aoquadrado = lambda x : x**2

Por exemplo: temos uma lista de tuplas e queremos ordenar a lista com base no segundo item da tupla. O método `sort` pode receber um argumento opcional `key` que nos diz como interpretar o item da lista para a ordenação.

In [92]:
pairs = [(1, 'um'), (2, 'dois'), (3, 'três'), (4, 'quatro')]
pairs.sort()
pairs

[(1, 'um'), (2, 'dois'), (3, 'três'), (4, 'quatro')]

In [93]:
pairs.sort(key=lambda p: p[1])
pairs

[(2, 'dois'), (4, 'quatro'), (3, 'três'), (1, 'um')]

Aqui usamos uma lambda em um extrato de uma lista (com a função filter).

In [145]:
list(filter(lambda x:x==1, [1,2,3]))

[1]

In [94]:
squares = [x**2 for x in range(100)]
sq = list(filter(lambda x : x%2 == 0 and x%3 == 0, squares))
sq

[0,
 36,
 144,
 324,
 576,
 900,
 1296,
 1764,
 2304,
 2916,
 3600,
 4356,
 5184,
 6084,
 7056,
 8100,
 9216]

# Miscellaneous
There are a lot of topics that we didn’t cover, as well as a lot of the python standard library that we won’t address. Here we introduce a few more concepts.

# Exceptions

Python raises exceptions when it encounters an error.  The idea is that you can trap these exceptions and take an appropriate action instead of causing the code to crash.  The mechanism for this is `try` / `except`.  Here's an example that causes an exception, `ZeroDivisionError`:

In [175]:
a = 1/0

ZeroDivisionError: division by zero

and here we handle this  

In [176]:
try:
    a = 1/0
except ZeroDivisionError:
    print("warning: you divided by zero")
    a = 1

a

1

In [177]:
def rec(a):
    try:
        b=1/a
    except ZeroDivisionError:
        print("warning: you divided by zero")
        b=None
    return b

In [178]:
print(rec(0))

None


Another example&mdash;trying to access a key that doesn't exist in a dictionary:

In [179]:
dict = {"a":1, "b":2, "c":3}
print(dict["d"])


KeyError: 'd'

`KeyError` is the exception that was raised.  We can check for this and take the appropriate action instead

In [181]:
try:
    val = dict["d"]
except KeyError:
    val = None

print(val)

None


In [185]:
a=[1,2,3]

for x in a:
    print(x)
    if x==3:
        raise ZeroDivisionError

1
2
3


ZeroDivisionError: 

There are a lot of different types of exceptions that you can catch, and you can catch multiple ones per except clause or have multiple except clauses.  You probably won't be able to anticipate every failure mode in advance.  In that case, when you run and your code crashes because of an exception, the python interpreter will print out the name of the exception and you can then modify your code to take the appropriate action.

# File I/O

One of the main things that we want to do in scientific computing is get data into and out of our programs.  In addition to plain text files, there are modules that can read lots of different data formats we might encounter.

As expected, a file is an object.  Here we'll use the `try`, `except` block to capture exceptions (like if the file cannot be opened). 

In [190]:
try: f = open("./mywrite.txt", "w")   # open for writing -- any file of the same name will be overwritten
except: 
    print("cannot open the file")

print(f)


<_io.TextIOWrapper name='./mywrite.txt' mode='w' encoding='UTF-8'>


In [191]:
f.write("this is my first write\n")
f.close()

We can easily loop over the lines in a file

In [193]:
try: 
    f = open("./test.txt", "r")
except:
    print("error: cannot open the file")
    
for line in f:
    print(line.split())
    
f.close()

['Lorem', 'ipsum', 'dolor', 'sit', 'amet,', 'consectetur', 'adipisicing', 'elit,', 'sed', 'do']
['eiusmod', 'tempor', 'incididunt', 'ut', 'labore', 'et', 'dolore', 'magna', 'aliqua.', 'Ut', 'enim', 'ad']
['minim', 'veniam,', 'quis', 'nostrud', 'exercitation', 'ullamco', 'laboris', 'nisi', 'ut']
['aliquip', 'ex', 'ea', 'commodo', 'consequat.', 'Duis', 'aute', 'irure', 'dolor', 'in']
['reprehenderit', 'in', 'voluptate', 'velit', 'esse', 'cillum', 'dolore', 'eu', 'fugiat', 'nulla']
['pariatur.', 'Excepteur', 'sint', 'occaecat', 'cupidatat', 'non', 'proident,', 'sunt', 'in']
['culpa', 'qui', 'officia', 'deserunt', 'mollit', 'anim', 'id', 'est', 'laborum.']
[]


As mentioned earlier, there are lots of string functions. Above we used `strip()` to remove the trailing whitespace and returns

Observe que se tentarmos abrir um arquivo que não existe, ele irá partir para o bloco do `except` e executará o `print`

In [194]:
try: 
    f = open("./tessasssst.txt", "r")
except:
    print("error: cannot open the file")

error: cannot open the file


In practice, we never read files explicitly but use some library. Specific data format ofen have dedicated modules, like `cvs` for cvs files and `configparser` for ini files. Nesta nota de aula iremos abordar apenas arquivos `cvs`.

comma-separated values are an easy way to exchange data – you can generate these from a spreadsheet program. In the example below, we are assuming that the first line of the spreadsheet/csv file gives the headings that identify the columns.

Note that there is an amazing amount of variation in terms of what can be in a CSV file and what the format is – the csv module does a good job sorting this all out for you.

Colocar algum exemplo lendo csv e depois ver algo para salvar em csv.

In [199]:
f = lambda x:x*2

a =[1,2,3]

list(map(f,a))

[x**2 for x in a]

[1, 4, 9]

In [205]:
print(a)

for x in a:
    if x==1:
        continue
    
    print(x)
    

[1, 2, 3]
2
3


In [206]:
a=[1,2,3,4,5,5,5]

In [210]:
for x in a:
    print(x)
print("")
for x in set(a):    
    print(x)

1
2
3
4
5
5
5

1
2
3
4
5


# Exercises

Work on those you like the most. For the exam, preparing 2 of these is enough (or say 3 if you pick those that are very short).

## Q1: Machine precision

When talking about floating point, we discussed _machine epsilon_, $\epsilon$&mdash;this is the smallest number that when added to 1 is still different from 1.

We'll compute $\epsilon$ here:

  * Pick an initial guess for $\epsilon$ of `eps = 1`.  

  * Create a loop that checks whether `1 + eps` is different from `1`
  
  * Each loop iteration, cut the value of `eps` in half
  
What value of $\epsilon$ do you find?



## Q2: Iterations

### Part 1

To iterate over the tuples, where the _i_-th tuple contains the _i_-th elements of certain sequences, we can use `zip(*sequences)` function.

We will iterate over two lists, `names` and `age`, and print out the resulting tuples.

  * Start by initializing lists `names = ["Mary", "John", "Sarah"]` and `age = [21, 56, 98]`.
  
  * Iterate over the tuples containing a name and an age, the `zip(list1, list2)` function might be useful here.
  
  * Print out formatted strings of the type "*NAME is AGE years old*".
  

### Part 2

The function `enumerate(sequence)` returns tuples containing indices of objects in the sequence, and the objects. 

The `random` module provides tools for working with the random numbers. In particular, `random.randint(start, end)` generates a random number not smaller than `start`, and not bigger than `end`.

  * Generate a list of 10 random numbers from 0 to 9.
  
  * Using the `enumerate(random_list)` function, iterate over the tuples of random numbers and their indices, and print out *"Match: NUMBER and INDEX"* if the random number and its index in the list match.

## Q3: Books

Here is a list of book titles (from http://thegreatestbooks.org).  Loop through the list and capitalize each word in each title. 

In [25]:
titles = ["don quixote", 
          "in search of lost time", 
          "ulysses", 
          "the odyssey", 
          "war and peace", 
          "moby dick", 
          "the divine comedy", 
          "hamlet", 
          "the adventures of huckleberry finn", 
          "the great gatsby"]

## Q4: Word counts

Here's some text (the Gettysburg Address).  Our goal is to count how many times each word repeats.  We'll do a brute force method first, and then we'll look at ways to do it more efficiently (and compactly).

In [26]:
gettysburg_address = """
Four score and seven years ago our fathers brought forth on this continent, 
a new nation, conceived in Liberty, and dedicated to the proposition that 
all men are created equal.

Now we are engaged in a great civil war, testing whether that nation, or 
any nation so conceived and so dedicated, can long endure. We are met on
a great battle-field of that war. We have come to dedicate a portion of
that field, as a final resting place for those who here gave their lives
that that nation might live. It is altogether fitting and proper that we
should do this.

But, in a larger sense, we can not dedicate -- we can not consecrate -- we
can not hallow -- this ground. The brave men, living and dead, who struggled
here, have consecrated it, far above our poor power to add or detract.  The
world will little note, nor long remember what we say here, but it can never
forget what they did here. It is for us the living, rather, to be dedicated
here to the unfinished work which they who fought here have thus far so nobly
advanced. It is rather for us to be here dedicated to the great task remaining
before us -- that from these honored dead we take increased devotion to that
cause for which they gave the last full measure of devotion -- that we here
highly resolve that these dead shall not have died in vain -- that this
nation, under God, shall have a new birth of freedom -- and that government
of the people, by the people, for the people, shall not perish from the earth.
"""

We've already seen the `.split()` method will, by default, split by spaces, so it will split this into words, producing a list:

In [27]:
ga = gettysburg_address.split()

In [182]:
ga

['Four',
 'score',
 'and',
 'seven',
 'years',
 'ago',
 'our',
 'fathers',
 'brought',
 'forth',
 'on',
 'this',
 'continent,',
 'a',
 'new',
 'nation,',
 'conceived',
 'in',
 'Liberty,',
 'and',
 'dedicated',
 'to',
 'the',
 'proposition',
 'that',
 'all',
 'men',
 'are',
 'created',
 'equal.',
 'Now',
 'we',
 'are',
 'engaged',
 'in',
 'a',
 'great',
 'civil',
 'war,',
 'testing',
 'whether',
 'that',
 'nation,',
 'or',
 'any',
 'nation',
 'so',
 'conceived',
 'and',
 'so',
 'dedicated,',
 'can',
 'long',
 'endure.',
 'We',
 'are',
 'met',
 'on',
 'a',
 'great',
 'battle-field',
 'of',
 'that',
 'war.',
 'We',
 'have',
 'come',
 'to',
 'dedicate',
 'a',
 'portion',
 'of',
 'that',
 'field,',
 'as',
 'a',
 'final',
 'resting',
 'place',
 'for',
 'those',
 'who',
 'here',
 'gave',
 'their',
 'lives',
 'that',
 'that',
 'nation',
 'might',
 'live.',
 'It',
 'is',
 'altogether',
 'fitting',
 'and',
 'proper',
 'that',
 'we',
 'should',
 'do',
 'this.',
 'But,',
 'in',
 'a',
 'larger',
 '

Now, the next problem is that some of these still have punctuation.  In particular, we see "`.`", "`,`", and "`--`".

When considering a word, we can get rid of these by using the `replace()` method:

In [183]:
a = "end.,"
b = a.replace(".", "").replace(",", "")
b

'end'

Another problem is case&mdash;we want to count "but" and "But" as the same.  Strings have a `lower()` method that can be used to convert a string:

In [184]:
a = "But"
b = "but"
a == b

False

In [185]:
a.lower() == b.lower()

True

Recall that strings are immutable, so `replace()` produces a new string on output.

### Your task

Create a dictionary that uses the unique words as keys and has as a value the number of times that word appears.  

Write a loop over the words in the string (using our split version) and do the following:
  * remove any punctuation
  * convert to lowercase
  * test if the word is already a key in the dictionary (using the `in` operator)
     - if the key exists, increment the word count for that key
     - otherwise, add it to the dictionary with the appropriate count of `1`.

At the end, print out the words and a count of how many times they appear

### More compact way

We can actually do this a lot more compactly by using another list comprehensions and another python datatype called a set.  A set is a group of items, where each item is unique (e.g., no repetitions).

Here's a list comprehension that removes all the punctuation and converts to lower case:

In [186]:
words = [q.lower().replace(".", "").replace(",", "") for q in ga]

and by using the `set()` function, we turn the list into a set, removing any duplicates:

In [187]:
unique_words = set(words)

now we can loop over the unique words and use the `count` method of a list to find how many there are

In [188]:
count = {}
for uw in unique_words:
    count[uw] = words.count(uw)
    
count

{'of': 5,
 'ago': 1,
 'their': 1,
 'perish': 1,
 'civil': 1,
 'should': 1,
 'four': 1,
 'do': 1,
 'not': 5,
 'freedom': 1,
 'altogether': 1,
 'a': 7,
 'unfinished': 1,
 'measure': 1,
 'or': 2,
 'little': 1,
 'before': 1,
 'proposition': 1,
 'live': 1,
 'all': 1,
 'created': 1,
 'living': 2,
 'struggled': 1,
 'say': 1,
 'great': 3,
 'world': 1,
 'note': 1,
 'fitting': 1,
 'add': 1,
 'in': 4,
 'resolve': 1,
 'rather': 2,
 'but': 2,
 'nation': 5,
 'detract': 1,
 'forget': 1,
 'power': 1,
 'dedicate': 2,
 'nor': 1,
 'far': 2,
 'cause': 1,
 'did': 1,
 'people': 3,
 'new': 2,
 'earth': 1,
 'long': 2,
 'have': 5,
 'hallow': 1,
 'final': 1,
 'these': 2,
 'proper': 1,
 'testing': 1,
 'here': 8,
 'be': 2,
 'score': 1,
 'come': 1,
 'whether': 1,
 'now': 1,
 'liberty': 1,
 'endure': 1,
 'is': 3,
 'honored': 1,
 'dead': 3,
 'might': 1,
 'above': 1,
 'battle-field': 1,
 'government': 1,
 'portion': 1,
 'dedicated': 4,
 'any': 1,
 'increased': 1,
 'continent': 1,
 'can': 5,
 'larger': 1,
 'task': 1,


Even shorter -- we can use a dictionary comprehension, like a list comprehension

In [189]:
c = {uw: count[uw] for uw in unique_words}

In [190]:
c

{'of': 5,
 'ago': 1,
 'their': 1,
 'perish': 1,
 'civil': 1,
 'should': 1,
 'four': 1,
 'do': 1,
 'not': 5,
 'freedom': 1,
 'altogether': 1,
 'a': 7,
 'unfinished': 1,
 'measure': 1,
 'or': 2,
 'little': 1,
 'before': 1,
 'proposition': 1,
 'live': 1,
 'all': 1,
 'created': 1,
 'living': 2,
 'struggled': 1,
 'say': 1,
 'great': 3,
 'world': 1,
 'note': 1,
 'fitting': 1,
 'add': 1,
 'in': 4,
 'resolve': 1,
 'rather': 2,
 'but': 2,
 'nation': 5,
 'detract': 1,
 'forget': 1,
 'power': 1,
 'dedicate': 2,
 'nor': 1,
 'far': 2,
 'cause': 1,
 'did': 1,
 'people': 3,
 'new': 2,
 'earth': 1,
 'long': 2,
 'have': 5,
 'hallow': 1,
 'final': 1,
 'these': 2,
 'proper': 1,
 'testing': 1,
 'here': 8,
 'be': 2,
 'score': 1,
 'come': 1,
 'whether': 1,
 'now': 1,
 'liberty': 1,
 'endure': 1,
 'is': 3,
 'honored': 1,
 'dead': 3,
 'might': 1,
 'above': 1,
 'battle-field': 1,
 'government': 1,
 'portion': 1,
 'dedicated': 4,
 'any': 1,
 'increased': 1,
 'continent': 1,
 'can': 5,
 'larger': 1,
 'task': 1,


## Q5: Foxes and dogs

### Part 1. Short words

Let's practice functions.  Here's a simple function that takes a string and returns a list of all the 4 letter words:

In [32]:
def four_letter_words(message):
    words = message.split()
    four_letters = [w for w in words if len(w) == 4]
    return four_letters

In [33]:
message = "The quick brown fox jumps over the lazy dog"
print(four_letter_words(message))

['over', 'lazy']


Write a version of this function that takes a second argument, n, that is the word length we want to search for

### Part 2: Panagrams

A _panagram_ is a sentence that includes all 26 letters of the alphabet, e.g., "_The quick brown fox jumps over the lazy dog_."

Write a function that takes as an argument a sentence and returns `True` or `False`, indicating whether the sentence is a panagram.

## Q6: Catch errors

We want to safely convert a string into a float, int, or leave it as a string, depending on its contents.  As we've already seen, python provides `float()` and `int()` functions for this:

In [30]:
a = "2.0"
b = float(a)
print(b, type(b))

2.0 <class 'float'>


But these throw exceptions if the conversion is not possible

In [31]:
a = "this is a string"
b = float(a)

ValueError: could not convert string to float: 'this is a string'

In [195]:
a = "1.2345"
b = int(a)
print(b, type(b))

ValueError: invalid literal for int() with base 10: '1.2345'

In [196]:
b = float(a)
print(b, type(b))

1.2345 <class 'float'>


Notice that an int can be converted to a float, but if you convert a float to an int, you risk losing significant digits.  A string cannot be converted to either.

### Your task

Write a function, `convert_type(a)` that takes a string `a`, and converts it to a float if it is a number with a decimal point, an int if it is an integer, or leaves it as a string otherwise, and returns the result.  You'll want to use exceptions to prevent the code from aborting.

## Q7: Tic-tac-toe

Here we'll write a simple tic-tac-toe game that 2 players can play.  First we'll create a string that represents our game board:

In [197]:
board = """
 {s1:^3} | {s2:^3} | {s3:^3}
-----+-----+-----
 {s4:^3} | {s5:^3} | {s6:^3}
-----+-----+-----      123
 {s7:^3} | {s8:^3} | {s9:^3}       456
                       789  
"""

This board will look a little funny if we just print it&mdash;the spacing is set to look right when we replace the `{}` with `x` or `o`

In [198]:
print(board)


 {s1:^3} | {s2:^3} | {s3:^3}
-----+-----+-----
 {s4:^3} | {s5:^3} | {s6:^3}
-----+-----+-----      123
 {s7:^3} | {s8:^3} | {s9:^3}       456
                       789  



and well use a dictionary to denote the status of each square, "x", "o", or empty, ""

In [199]:
play = {}

def initialize_board(play):
    for n in range(9):
        play["s{}".format(n+1)] = ""

initialize_board(play)
play

{'s1': '',
 's2': '',
 's3': '',
 's4': '',
 's5': '',
 's6': '',
 's7': '',
 's8': '',
 's9': ''}

Note that our `{}` placeholders in the `board` string have identifiers (the numbers in the `{}`).  We can use these to match the variables we want to print to the placeholder in the string, regardless of the order in the `format()`

In [200]:
a = "{s1:} {s2:}".format(s2=1, s1=2)
a

'2 1'

Here's an easy way to add the values of our dictionary to the appropriate squares in our game board.  First note that each of the {} is labeled with a number that matches the keys in our dictionary.  Python provides a way to unpack a dictionary into labeled arguments, using **

This lets us to write a function to show the tic-tac-toe board.

In [201]:
def show_board(play):
    """ display the playing board.  We take a dictionary with the current state of the board
    We rely on the board string to be a global variable"""
    print(board.format(**play))
    
show_board(play)


     |     |    
-----+-----+-----
     |     |    
-----+-----+-----      123
     |     |           456
                       789  



Now we need a function that asks a player for a move:

In [202]:
def get_move(n, xo, play):
    """ ask the current player, n, to make a move -- make sure the square was not 
        already played.  xo is a string of the character (x or o) we will place in
        the desired square """
    valid_move = False
    while not valid_move:
        idx = input("player {}, enter your move (1-9)".format(n))
        if play["s{}".format(idx)] == "":
            valid_move = True
        else:
            print("invalid: {}".format(play["s{}".format(idx)]))
            
    play["s{}".format(idx)] = xo

In [203]:
help(get_move)

Help on function get_move in module __main__:

get_move(n, xo, play)
    ask the current player, n, to make a move -- make sure the square was not 
    already played.  xo is a string of the character (x or o) we will place in
    the desired square



### Your task

Using the functions defined above,
  * `initialize_board()`
  * `show_board()`
  * `get_move()`

fill in the function `play_game()` below to complete the game, asking for the moves one at a time, alternating between player 1 and 2

In [ ]:
def play_game():
    """ play a game of tic-tac-toe """
    
    play ={}
    initialize_board(play)
    show_board(play)

In [ ]:
play_game()

## Q8: Shopping cart

Let's write a simple shopping cart class -- this will hold items that you intend to purchase as well as the amount, etc.  And allow you to add / remove items, get a subtotal, etc.

We'll use two classes: `Item` will be a single item and `ShoppingCart` will be the collection of items you wish to purchase.

First, our store needs an inventory -- here's what we have for sale:

In [204]:
INVENTORY_TEXT = """
apple, 0.60
banana, 0.20
grapefruit, 0.75
grapes, 1.99
kiwi, 0.50
lemon, 0.20
lime, 0.25
mango, 1.50
papaya, 2.95
pineapple, 3.50
blueberries, 1.99
blackberries, 2.50
peach, 0.50
plum, 0.33
clementine, 0.25
cantaloupe, 3.25
pear, 1.25
quince, 0.45
orange, 0.60
"""

# this will be a global -- convention is all caps
INVENTORY = {}
for line in INVENTORY_TEXT.splitlines():
    if line.strip() == "":
        continue
    item, price = line.split(",")
    INVENTORY[item] = float(price)


In [205]:
INVENTORY

{'apple': 0.6,
 'banana': 0.2,
 'grapefruit': 0.75,
 'grapes': 1.99,
 'kiwi': 0.5,
 'lemon': 0.2,
 'lime': 0.25,
 'mango': 1.5,
 'papaya': 2.95,
 'pineapple': 3.5,
 'blueberries': 1.99,
 'blackberries': 2.5,
 'peach': 0.5,
 'plum': 0.33,
 'clementine': 0.25,
 'cantaloupe': 3.25,
 'pear': 1.25,
 'quince': 0.45,
 'orange': 0.6}

### Item 

Let's write an item class now -- we want it to hold the name and quantity.  

You should have the following features:

* The name should be something in our inventory

* Our shopping cart will include a list of all the items we want to buy, so we want to be able to check for duplicates.  Implement the equal test, `==`, using `__eq__`

* We'll want to consolidate duplicates, so implement the `+` operator, using `__add__` so we can add items together in our shopping cart.  Note, add should raise a ValueError if you try to add two `Items` that don't have the same name.

Here's a start:

In [206]:
class Item:
    """ an item to buy """
    
    def __init__(self, name, quantity=1):
        """keep track of an item that is in our inventory"""
        if name not in INVENTORY:
            raise ValueError("invalid item name")
        self.name = name
        self.quantity = quantity
        
    def __repr__(self):
        return "{}: {}".format(self.name, self.quantity)
        
    def __eq__(self, other):
        """check if the items have the same name"""
        return self.name == other.name
    
    def __add__(self, other):
        """add two items together if they are the same type"""
        if self.name == other.name:
            return Item(self.name, self.quantity + other.quantity)
        else:
            raise ValueError("names don't match")

Here are some tests your code should pass:

In [207]:
a = Item("apple", 10)
b = Item("banana", 20)

In [208]:
c = Item("apple", 20)

In [209]:
# won't work
a + b

ValueError: names don't match

In [210]:
# will work
a += c
print(a)

apple: 30


In [211]:
d = Item("dog")

ValueError: invalid item name

In [212]:
# should be False
a == b

False

In [213]:
# should be True -- they have the same name
a == c

True

How do they behave in a list?

In [214]:
items = []
items.append(a)
items.append(b)
items

[apple: 30, banana: 20]

In [215]:
# should be True -- they have the same name
c in items

True

### ShoppingCart

Now we want to create a shopping cart.  The main thing it will do is hold a list of items.

In [216]:
class ShoppingCart:
    
    def __init__(self):
        # the list of items we control
        self.items = []
        
    def subtotal(self):
        """ return a subtotal of our items """
        pass

    def add(self, name, quantity):
        """ add an item to our cart -- the an item of the same name already
        exists, then increment the quantity.  Otherwise, add a new item
        to the cart with the desired quantity."""
        pass
        
    def remove(self, name):
        """ remove all of item name from the cart """
        pass
        
    def report(self):
        """ print a summary of the cart """
        for item in self.items:
            print(f"{item.name} : {item.quantity}")

Here are some tests

In [217]:
sc = ShoppingCart()
sc.add("orange", 19)

In [218]:
sc.add("apple", 2)

In [219]:
sc.report()

In [220]:
sc.add("apple", 9)

In [221]:
# apple should only be listed once in the report, with a quantity of 11
sc.report()

In [222]:
sc.subtotal()

In [223]:
sc.remove("apple")

In [224]:
# apple should no longer be listed
sc.report()

## Q9: Poker odds

Use the deck of cards class from the notebook we worked through class to write a _Monte Carlo_ code that plays a lot of hands of straight poker (like 100,000).  Count how many of these hands has a particular poker hand (like 3-of-a-kind).  The ratio of # of hands with 3-of-a-kind to total hands is an approximation to the odds of getting a 3-of-a-kind in poker.

### Bonus: 
Just to practice modules, write that into a `.py` file to allow you to import and reuse them here.

## Q10: Tic-Tac-Toe again

Revisit the tic-tac-toe game you developed in the functions exercises but now write it as a class with methods to do each of the main steps.  

## Q11: Rock-Paper-Scissors

Implement a set of games of rock-paper-scissors against the computer.  

  * Ask for input from the user ("rock", "paper", or "scissors") and the randomly select one of these for the computer's play.
  * Announce who won.
  * Keep playing until a player says that they no longer want to play.
  * When all games are done, print out how many games were won by the player and by the computer 

## Q12: Pascal's triangle

Pascal's triangle is created such that each layer has 1 more element than the previous, with `1`s on the side and in the interior, the numbers are the sum of the two above it, e.g.,:
```
            1
          1   1
        1   2   1
      1   3   3   1
    1   4   6   4   1
  1   5   10  10  5   1
```

1. Write a function to return the first `n` rows of Pascal's triangle.  The return should be a list of length `n`, with each element itself a list containing the numbers for that row.
2. Write a function to print out Pascal's triangle with proper formatting, so the numbers in each row are centered between the ones in the row above

## Q13: Calendar events

We want to keep a schedule of events.  We will do this by creating a class called `Day`.  It is sketched out below.  A `Day` holds a list of events and has methods that allow you to add an delete events.  Our events will be instances of a class `Event`, which holds the time, location, and description of the event.

Finally, we can keep track of a list of all the `Day`s for which we have events to make our schedule.

Fill in these classes and write some code to demonstrate their use:

  * Create a full week of days in your calendar
  * Add an event every day at noon called "lunch"
  * Randomly add some other events to fill out your calendar
  * Write some code that tells you the start time of your first meeting and the end time of your last meeting (this is the length of your work day)

In [225]:
class Day:
    """a single day keeping track of the events scheduled"""
    def __init__(self, month, day, year):
        # store the month, day, and year as data in the class
        
        # keep track of the events
        self.events = []
    
    def add_event(self, name, time=None, location=None):
        pass
    
    def delete_event(self, name):
        pass
    
    
class Event:
    """a single event in our calendar"""
    def __init__(self, name, time=9, location=None, duration=1):
        self.name = name
        self.time = time
        self.location = location
        self.duration = duration